server.cpp

```diff
@@ -134,46 +156,141 @@ void handle_request(struct server_app *app, int client_socket) {
-    // The following code returns a dummy response for all requests
-    // but it should give you a rough idea about what a proper response looks like
-    // What you need to do 
-    // (when the requested file exists):
-    // * Open the requested file
-    // * Build proper response headers (see details in the spec), and send them
-    // * Also send file content
-    // (When the requested file does not exist):
-    // * Generate a correct response
-
-    char response[] = "HTTP/1.0 200 OK\r\n"
-                      "Content-Type: text/plain; charset=UTF-8\r\n"
-                      "Content-Length: 15\r\n"
-                      "\r\n"
-                      "Sample response";
+void serve_local_file(int client_socket, std::string path) {
+    // build response using a stream
+    std::stringstream response_stream;
 
-    send(client_socket, response, strlen(response), 0);
+    // Per piazza @71, all file paths start with a /
+    // so remove the first character
+    path = path.substr(1);
+
+    // search for path in current directory
+    // NOTE: per the spec, we are not recursively searching cur dir
+    const fs::path cur_dir = fs::current_path();
+    bool file_found = false;
+
+    for (const auto& entry : fs::directory_iterator(cur_dir)) {
+        if (entry.is_regular_file() && entry.path().filename() == path) {
+            file_found = true;
+            break;
+        }
+    }
+
+    if (file_found) {
+        std::ifstream file(path);
+
+        if (file.is_open()) {
+            // read the file as bytes
+            std::vector<char> file_buf;
+            char cur_byte;
+
+            while (file.get(cur_byte)) {
+                file_buf.push_back(cur_byte);
+            }
+
+            size_t content_length = file_buf.size();
+
+            // extract content type
+            // reference: https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types
+            filetype ftype = determine_filetype(path);
+            std::string content_type;
+            switch (ftype) {
+                case ft_html:
+                    content_type = "text/html; charset=UTF-8";
+                    break;
+                case ft_jpg:
+                    content_type = "image/jpeg";
+                    break;
+                case ft_txt:
+                    content_type = "text/plain; charset=UTF-8";
+                    break;
+                default:
+                    content_type = "application/octet-stream";
+            }
+
+            response_stream << "HTTP/1.0 200 OK\r\n";
+            response_stream << "Date: " << current_http_date() << "\r\n";
+            response_stream << "Connection: close\r\n";
+            response_stream << "Content-Type: " << content_type << "\r\n";
+            response_stream << "Content-Length: " << content_length << "\r\n";
+            // CRLF before entity body
+            response_stream << "\r\n";
+
+            // entity body
```

Comment:
Should note that we need to use .write in this way, rather than << because the entity body may contain null bytes, which would just terminate a string stream